src/librustc/middle/resolve_lifetime.rs

```diff
@@ -299,22 +316,92 @@ type ScopeRef<'a> = &'a Scope<'a>;
 
 const ROOT_SCOPE: ScopeRef<'static> = &Scope::Root;
 
-pub fn krate(
-    sess: &Session,
-    cstore: &CrateStore,
-    hir_map: &Map,
-) -> Result<NamedRegionMap, ErrorReported> {
-    let krate = hir_map.krate();
+pub fn provide(providers: &mut ty::maps::Providers) {
+    *providers = ty::maps::Providers {
+        resolve_lifetimes,
+
+        named_region_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE).defs.get(&id).cloned()
+        },
+
+        is_late_bound_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE)
+                .late_bound
+                .get(&id)
+                .cloned()
+        },
+
+        object_lifetime_defaults_map: |tcx, id| {
+            let id = LocalDefId::from_def_id(DefId::local(id)); // (*)
+            tcx.resolve_lifetimes(LOCAL_CRATE)
+                .object_lifetime_defaults
+                .get(&id)
+                .cloned()
+        },
+
+        ..*providers
+    };
+
+    // (*) FIXME the query should be defined to take a LocalDefId
+}
+
+/// Computes the `ResolveLifetimes` map that contains data for the
+/// entire crate. You should not read the result of this query
+/// directly, but rather use `named_region_map`, `is_late_bound_map`,
+/// etc.
+fn resolve_lifetimes<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    for_krate: CrateNum,
+) -> Rc<ResolveLifetimes> {
+    assert_eq!(for_krate, LOCAL_CRATE);
+
+    let named_region_map = krate(tcx).unwrap_or_default();
```

Comment:
This looks a bit violent and likely to cause cascade errors - if resolving lifetimes for *one* item fails, it will destroy all lifetime information without aborting compilation.